crates/pico/src/memo_ref.rs

```diff
@@ -2,19 +2,36 @@ use std::{marker::PhantomData, ops::Deref};
 
 use intern::InternId;
 
-use crate::{Database, DerivedNodeId, ParamId};
+use crate::{DatabaseDyn, DerivedNodeId, ParamId};
 
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[derive(Debug)]
 pub struct MemoRef<T> {
-    pub(crate) db: *const Database,
+    pub(crate) db: *const dyn DatabaseDyn,
     pub(crate) derived_node_id: DerivedNodeId,
     phantom: PhantomData<T>,
 }
 
+impl<T> Copy for MemoRef<T> {}
+
+impl<T> Clone for MemoRef<T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<T> PartialEq for MemoRef<T> {
+    fn eq(&self, other: &Self) -> bool {
+        std::ptr::eq(self.db, other.db) && self.derived_node_id == other.derived_node_id
+    }
+}
+
+impl<T> Eq for MemoRef<T> {}
+
+#[allow(clippy::unnecessary_cast)]
 impl<T: 'static + Clone> MemoRef<T> {
-    pub fn new(db: &Database, derived_node_id: DerivedNodeId) -> Self {
+    pub fn new(db: &dyn DatabaseDyn, derived_node_id: DerivedNodeId) -> Self {
         Self {
-            db,
+            db: db as *const _ as *const dyn DatabaseDyn,
```

Comment:
This is suspicious. Clippy says that it's unnecessary cast, but without it we got lifetime issues